File: Dynamic_Programming/3544. Subtree Inversion Sum.cpp
// TC:O(n * k) SC:O(n * k)
class Solution {
public:
    long long subtreeInversionSum(vector<vector<int>>& edges, vector<int>& nums, int k) {
        int n = nums.size();
        vector<vector<int>> g(n);
        for (auto& e : edges) {
            int x = e[0], y = e[1];
            g[x].push_back(y);
            g[y].push_back(x);
        }

        vector memo(n, vector<array<long long, 2>>(k, {LLONG_MIN, LLONG_MIN}));
        auto dfs = [&](this auto&& dfs, int x, int fa, int cd, bool parity) -> long long {
            auto& res = memo[x][cd][parity];
            if (res != LLONG_MIN) {
                return res;
            }
            res = parity ? -nums[x] : nums[x];
            for (int y : g[x]) {
                if (y != fa) {
                    res += dfs(y, x, max(cd - 1, 0), parity);
                }
            }
            if (cd == 0) {
                long long s = parity ? nums[x] : -nums[x];
                for (int y : g[x]) {
                    if (y != fa) {
                        s += dfs(y, x, k - 1, !parity);
                    }
                }
                res = max(res, s);
            }

            return res;
        };

        return dfs(0, -1, 0, 0);
    }
};

File: Dynamic_Programming/3544. Subtree Inversion Sum.py
# TC:O(n * k) SC:O(n * k)
class Solution:
    def subtreeInversionSum(self, edges: List[List[int]], nums: List[int], k: int) -> int:
        graph = [[] for _ in range(len(nums))]
        for e in edges:
            graph[e[0]].append(e[1])
            graph[e[1]].append(e[0])
        memo = {}
        def dfs(u:int, fa:int, cd:int, mul:int) -> int:
            t = (u, cd, mul)
            if t in memo:
                return memo[t]
            ans = nums[u] * mul
            # 不翻轉
            for nxt in graph[u]:
                if nxt != fa:
                    ans += dfs(nxt, u, cd - 1 if cd else 0, mul)
            if cd == 0: # 可翻轉
                ret = nums[u] * mul * -1
                for nxt in graph[u]:
                    if nxt != fa:
                        ret += dfs(nxt, u, k - 1, mul * -1)
                if ret > ans:
                    ans = ret
            memo[t] = ans
            return ans
        return dfs(0, -1, 0, 1)
        
